### ISSUE TITLE
2.1.4版在Firefox Android上option.html出現空白頁因而未能啟用

### ISSUE THREAD

**Issue by SakuraiNaomi** [tongwentang/tongwentang-extension#48]
Firefox Android 68.11
Firefox Android Nightly 83.0a1 和 97.0a1
都相同情况，頁面只看見半秒就轉成全白不能選自動轉換
68.11最後一版不像Nighly加Collection必定只能加入最新2.1.4
放回同文1.5頁面就正常出來了

有爬過一會Closed issue知道作者繁忙，但是68已經有好些網站
不能正常運作，也花了一周沒找到好的替代方案... 祝聖誕快樂 :D

![Screenshot_20211220-125907](https://user-images.githubusercontent.com/96416439/147094610-ce312ad0-e520-4b43-8f67-85ed3c676293.jpg)

**Comment by t7yang** [comment]
是說，如果用 1.5 版本就完全可以正常運作嗎？

**Comment by SakuraiNaomi** [comment]
因為我只有用自動轉換，bbs網站上的簡體字的確有轉換成繁體字，
其他功能不敢保證正常，68.11裝上1.5同一頁面這樣子才能啓動

![Screenshot_20220207-192930](https://user-images.githubusercontent.com/96416439/152773491-77a0cd0e-60c7-479b-a580-d58c202a43b8.jpg)

**Comment by t7yang** [comment]
你的 BBS 是指 term.ptt 還是純網頁。因為敘述看起來我不是很容易理解。
或是方便的話，請你稍微一小段影片展示兩個版本之間的差異，這樣比較觀察你碰到的是什麼問題。
目前 Github 已經支援上傳影片。

**Comment by SakuraiNaomi** [comment]
3dots > Addon > Addon Manager > 新同文堂 > 設定
2.1.4版白色頁面無法選擇自動轉換以啟動同文堂功能
1.5版正常

https://user-images.githubusercontent.com/96416439/156893372-5aeefe4a-a1d5-4f90-97a5-b4816be0e146.mp4

**Comment by t7yang** [comment]
不確實是不是用了新語法導致錯誤，還需要再詳細研究。

**Comment by htbkoo** [comment]
(不好意思, 在 GitHub 上面我比較習慣使用英文 - 如果有寫不清楚的地方, 我可以再用中文說明)

---

TL;DR: I tried to fix the extension with these changes - not sure if this are the only changes we need, but at least the option page is fixed, and the auto-translation (from simplified Chinese to traditional Chinese) now works for me:

https://github.com/htbkoo/tongwentang-extension/compare/master...htbkoo:tongwentang-extension:fix-android-fenix

---

More details:

I had tried to debug the extension with `Firefox Nightly` based on [this instruction](https://extensionworkshop.com/documentation/develop/developing-extensions-for-firefox-for-android/) and I had found that there are few errors preventing the extension from working on `Android` 

I. There is an error `Uncaught TypeError: can't access property "autoConvert", n is undefined`. when opening the option page

After some investigation, I think the error is thrown from here ([`GeneralSettings.tsx#L16`](https://github.com/tongwentang/tongwentang-extension/blob/master/src/options/pages/general/GeneralSettings.tsx#L16)) and `general` is `undefined`.

I am suspecting that the `general` object got overwritten to `undefined` here ([`use-general-opt.ts#L38`](https://github.com/tongwentang/tongwentang-extension/blob/master/src/options/hooks/general/use-general-opt.ts#L38)) when the effect hook try to read the value from storage when the storage is empty (please correct me if I am wrong).

I think we can probably fix this by either:
1. using optional chaining at `<GeneralSettings />` (please see this for details: https://github.com/htbkoo/tongwentang-extension/compare/master...htbkoo:tongwentang-extension:fix-android-fenix-check-before-setting-value?expand=1)
2. avoid setting the value to `general` at `use-general-opt.ts` if the value is `undefined` (please see this for details: https://github.com/htbkoo/tongwentang-extension/compare/master...htbkoo:tongwentang-extension:fix-android-fenix-optional-chaining)

II. The `Context Menu` option page and `Word` option page failed to load

Unfortunately, the `Context Menu` option page and `Word` option page also failed to load. 

Similarly, I had tried to fix them by avoid setting the value if the value is `undefined` (Please see this commit for details: https://github.com/htbkoo/tongwentang-extension/commit/e3bf1e5b116837f356cb5b89efb153ff8a35c3ce)

III. Background script failed to run due to the fact that some functions are unsupported on Android

After applying the above fixes, the option page would work but the translation still failed to run. 

And I found that it is because the background script did not successfully run because some of the features (e.g. `contextMenus`; `commands`) are unsupported on `Android` `Firefox` and some `functions` are `undefined` (e.g. `commands.onCommand`; `browser.contextMenus.create`; `browser.contextMenus.remove`). 

FYI, this is also hinted by the warnings by `Firefox` that some of the fields in [`manifest.json` (e.g. `permissions`.`contextMenus`; `commands`) are unsupported](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Browser_compatibility_for_manifest.json)

I ended up trying to fix the issue by adding optional chaining for all offending usages. (Please see this for details: https://github.com/htbkoo/tongwentang-extension/compare/fix-android-fenix-other-options-pages-error...htbkoo:tongwentang-extension:fix-android-fenix-background-script)

IV. `tabs.detectLanguage` is not a function

This is quite strange but somehow `tabs.detectLanguage` is undefined on my `Firefox` and that was the last obstacle that blocked the extension from working. 

In order to fix the issue, I added a `try-catch` around the usage and similarly defaulted the language to `ZhType.und` when there is any error caught (Please refer to this commit for details: https://github.com/htbkoo/tongwentang-extension/commit/973f7d7464bb8e00589c98e04500a80f03487f57)

---

Please let me know if you need more information.

Thank you!

---

Environment:
Extension version: `New TongWenTang 2.1.6`
Browser version: `Firefox Nightly (107.0a1)`
Device: `Google Pixel 6`
OS: `Android 13`

**Comment by htbkoo** [comment]
Just another small updates for your information:

I ended up switching to use the [Kiwi Browser](https://kiwibrowser.com/) ([kiwibrowser/src](https://github.com/kiwibrowser/src)) instead.

And the extension just works out of the box (note that [Kiwi Broswer provides native support for extensions](https://kiwibrowser.com/features/)).

So I guess the issues are exclusive for `Fenix` on `Android`. 🤔

**Comment by null-von-sushi** [comment]
Has there been any progress on this? I'd love to switch to Firefox but am stuck on Kiwi as it's the only way for me to use this extension. 

Even if the extension settings page isn't fixed to work on mobile, I would still be happy to at least know if a workaround to set the auto-convert mode on

**Comment by t7yang** [comment]
hi @null-von-sushi 
the current target is MV3 migration.
will take a look after the migration.

**Comment by null-von-sushi** [comment]
In the meantime, is it possible to install the 1.5 version on current Firefox (Android)? I cannot seem to find the 1.5 version anywhere
